Expose linkLocalResourceResponse to _WKContextMenuElementInfo

arenevier opened the pull request:
#### 98458807c5ab6ff9d3b571987f03c66edf3aca2a
<pre>
Expose linkLocalResourceResponse to _WKContextMenuElementInfo
<a href="https://bugs.webkit.org/show_bug.cgi?id=293166">https://bugs.webkit.org/show_bug.cgi?id=293166</a>
<a href="https://rdar.apple.com/122194886">rdar://122194886</a>

Reviewed by Alex Christensen.

This will allow the UI process to get that information about a link
local data when a context menu is shown.

If the context menu is over a link, and there is local data available
for that url, linkLocalResourceResponse will reference an NSURLResponse
associated with that local data.

It will give access, among other things, to the data suggested file
name, and to the data mime type.

* Source/WebKit/Shared/API/Cocoa/_WKHitTestResult.h:
* Source/WebKit/Shared/API/Cocoa/_WKHitTestResult.mm:
(-[_WKHitTestResult linkLocalDataMIMEType]):
(-[_WKHitTestResult linkLocalDataSuggestedFilename]):
* Source/WebKit/Shared/WebHitTestResultData.cpp:
(WebKit::linkLocalDataSuggestedFilenameFromHitTestResult):
(WebKit::WebHitTestResultData::WebHitTestResultData):
* Source/WebKit/Shared/WebHitTestResultData.h:
* Source/WebKit/Shared/WebHitTestResultData.serialization.in:
* Source/WebKit/UIProcess/API/APIHitTestResult.h:
(API::HitTestResult::linkLocalDataSuggestedFilename const):
* Tools/TestWebKitAPI/Tests/mac/ContextMenuTests.mm:
(TestWebKitAPI::TEST(ContextMenuTests, HitTestResultLinkLocalDataSuggestedFilename)):

Canonical link: <a href="https://commits.webkit.org/295183@main">https://commits.webkit.org/295183@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/278d327905e77c0a7469be925158d245dc6ccb83

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104297 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24001 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14414 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109498 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54965 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/106337 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24373 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32548 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79206 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107303 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18942 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94121 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59534 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/18740 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12189 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54325 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/88446 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12247 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111879 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31454 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23175 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88231 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31818 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90361 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87894 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32790 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10555 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/25922 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16939 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31382 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36697 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31176 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34512 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32736 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/Shared/WebHitTestResultData.cpp:
```diff
@@ -114,6 +126,7 @@ WebHitTestResultData::WebHitTestResultData(const HitTestResult& hitTestResult, c
 
     elementType = elementTypeFromHitTestResult(hitTestResult);
     linkLocalDataMIMEType = linkLocalDataMIMETypeFromHitTestResult(hitTestResult);
+    linkLocalDataSuggestedFilename = linkLocalDataSuggestedFilenameFromHitTestResult(hitTestResult);
```
I think these should all be moved to the initializers above and out of the constructor body.

arenevier replied:
Done. I also refactored the constructors to remove duplication of that long initializer list.

achristensen07 commented on Tools/TestWebKitAPI/Tests/mac/ContextMenuTests.mm:
```diff
@@ -767,6 +767,27 @@ static void swizzledOrderFrontColorPanel(id, SEL, id)
+    __block bool gotProposedMenu = false;
+    [delegate setGetContextMenuFromProposedMenu:^(NSMenu *menu, _WKContextMenuElementInfo *elementInfo, id<NSSecureCoding>, void (^completion)(NSMenu *)) {
+        EXPECT_NOT_NULL(elementInfo.hitTestResult);
+        EXPECT_TRUE([elementInfo.hitTestResult.linkLocalDataSuggestedFilename isEqualToString:@"sunset-in-cupertino-600px.jpg"]);
```
EXPECT_WK_STREQ can be used to compare NSStrings to c-strings.  That way if there is a difference it will print out the strings instead of just "false"

arenevier replied:
done

achristensen07 commented on Tools/TestWebKitAPI/Tests/mac/ContextMenuTests.mm:
```diff
@@ -767,6 +767,27 @@ static void swizzledOrderFrontColorPanel(id, SEL, id)
+    [webView mouseDownAtPoint:NSMakePoint(200, 200) simulatePressure:NO withFlags:0 eventType:NSEventTypeRightMouseDown];
+    [webView mouseUpAtPoint:NSMakePoint(200, 200) withFlags:0 eventType:NSEventTypeRightMouseUp];
+    Util::run(&gotProposedMenu);
+}
```
It would be good to add to the test a case where it is null, such as if it right clicks on nothing.

arenevier replied:
done

achristensen07 commented on Source/WebKit/Shared/API/Cocoa/_WKHitTestResult.h:
```diff
@@ -47,6 +47,7 @@ WK_CLASS_AVAILABLE(macos(10.12), ios(16.0))
 @property (nonatomic, readonly, copy) NSURL *absoluteLinkURL;
 @property (nonatomic, readonly) BOOL hasLocalDataForLinkURL WK_API_AVAILABLE(macos(15.0), ios(18.0), visionos(2.0));
 @property (nonatomic, readonly, copy) NSString *linkLocalDataMIMEType WK_API_AVAILABLE(macos(15.0), ios(18.0), visionos(2.0));
```
I think we might not need to do this PR at all, actually.  The nullness of linkLocalDataMIMEType already exposes whether local data exists, and the suggested filename can be derived from absoluteImageURL by putting it into an NSURLRequest and accessing its suggestedFilename in the UI process like we do in ResourceResponse::platformSuggestedFilename.

achristensen07 replied:
Actually, that might lose some information from the response's Content-Disposition or Content-Type header fields.  Let's keep this approach.

achristensen07 replied:
Actually, let's expose a nullable NSURLResponse.  That'll have all the info needed for all the LocalData interfaces: hasLocalDataForLinkURL, linkLocalDataMIMEType, and the new one.

arenevier replied:
done

achristensen07 commented on Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
```diff
@@ -1510,4 +1510,25 @@ void WebFrame::setAppBadge(const WebCore::SecurityOriginData& origin, std::optio
+        return loader->response();
+
+    // Next, try subresources.
+    RefPtr<ArchiveResource> resource = loader->subresource(url);
```
I don't think these comments add any additional understanding to anyone reading the code.  Let's omit them.
Also, deduction guides can make it so you don't need to explicitly state the type as a template parameter, and the scope of this can be in the if statement:
if (RefPtr resource = loader->subresource(url))

achristensen07 commented on Source/WebKit/UIProcess/API/APIHitTestResult.h:
```diff
@@ -90,6 +90,10 @@ class HitTestResult : public API::ObjectImpl<API::Object::Type::HitTestResult> {
 
     bool allowsFollowingImageURL() const { return m_data.allowsFollowingImageURL; }
 
+    const std::optional<WebCore::ResourceResponse>& linkLocalResourceResponse() const {
```
Either everything should be on the same line (which would be better in this case) or the { needs to be on the next line.

achristensen07 commented on Source/WebKit/Shared/API/Cocoa/_WKHitTestResult.mm:
```diff
@@ -150,6 +150,14 @@ - (_WKHitTestResultElementType)elementType
+- (NSURLResponse *)linkLocalResourceResponse
+{
+    if (auto response = _hitTestResult->linkLocalResourceResponse()) {
+        return [response->nsURLResponse() copy];
```
This is a memory leak because copy returns a retained object.  In this case I think the best solution is just to return response->nsURLResponse() and annotate it as (nonatomic, readonly) in the header instead of (nonatomic, readonly, copy)

achristensen07 commented on Source/WebKit/Shared/API/Cocoa/_WKHitTestResult.mm:
```diff
@@ -150,6 +150,14 @@ - (_WKHitTestResultElementType)elementType
 
+- (NSURLResponse *)linkLocalResourceResponse
+{
+    if (auto response = _hitTestResult->linkLocalResourceResponse()) {
```
Omit { } for one line if statements.

achristensen07 commented on Source/WebKit/Shared/WebHitTestResultData.cpp:
```diff
@@ -178,7 +156,13 @@ WebHitTestResultData::WebHitTestResultData(const HitTestResult& hitTestResult, b
 
-WebHitTestResultData::WebHitTestResultData(const String& absoluteImageURL, const String& absolutePDFURL, const String& absoluteLinkURL, const String& absoluteMediaURL, const String& linkLabel, const String& linkTitle, const String& linkSuggestedFilename, const String& imageSuggestedFilename, bool isContentEditable, const WebCore::IntRect& elementBoundingBox, const WebKit::WebHitTestResultData::Is ...
+WebHitTestResultData::WebHitTestResultData(const HitTestResult& hitTestResult, const String& toolTipText)
+    : WebHitTestResultData(hitTestResult, toolTipText, false) {}
```
I'm not sure why this is changed, but let's try to minimize changes that aren't adding a response in this PR.  It does look like it could be cleaned up a bit, but in a separate PR.

achristensen07 commented on Source/WebKit/Shared/WebHitTestResultData.h:
```diff
@@ -132,7 +135,8 @@ struct WebHitTestResultData {
     std::optional<WebCore::SharedMemory::Handle> getImageSharedMemoryHandle() const;
-
+private:
+   WebHitTestResultData(const WebCore::HitTestResult& hitTestResult, const String& toolTipText, bool includeImage);
```
Why is this added?

arenevier committed:
Expose linkLocalResourceResponse to _WKContextMenuElementInfo
https://bugs.webkit.org/show_bug.cgi?id=293166
rdar://122194886

Reviewed by Alex Christensen.

This will allow the UI process to get that information about a link
local data when a context menu is shown.

If the context menu is over a link, and there is local data available
for that url, linkLocalResourceResponse will reference an NSURLResponse
associated with that local data.

It will give access, among other things, to the data suggested file
name, and to the data mime type.

* Source/WebKit/Shared/API/Cocoa/_WKHitTestResult.h:
* Source/WebKit/Shared/API/Cocoa/_WKHitTestResult.mm:
(-[_WKHitTestResult linkLocalDataMIMEType]):
(-[_WKHitTestResult linkLocalDataSuggestedFilename]):
* Source/WebKit/Shared/WebHitTestResultData.cpp:
(WebKit::linkLocalDataSuggestedFilenameFromHitTestResult):
(WebKit::WebHitTestResultData::WebHitTestResultData):
* Source/WebKit/Shared/WebHitTestResultData.h:
* Source/WebKit/Shared/WebHitTestResultData.serialization.in:
* Source/WebKit/UIProcess/API/APIHitTestResult.h:
(API::HitTestResult::linkLocalDataSuggestedFilename const):
* Tools/TestWebKitAPI/Tests/mac/ContextMenuTests.mm:
(TestWebKitAPI::TEST(ContextMenuTests, HitTestResultLinkLocalDataSuggestedFilename)):

Canonical link: https://commits.webkit.org/295183@main
